src/main/java/br/com/ks/viagem/emvoos/controllers/CompanhiaController.java

```diff
@@ -0,0 +1,58 @@
+package br.com.ks.viagem.emvoos.controllers;
+
+
+import br.com.ks.viagem.emvoos.DTOs.request.CompanhiaRequest;
+import br.com.ks.viagem.emvoos.DTOs.response.CompanhiaResponse;
+import br.com.ks.viagem.emvoos.exception.ResourceNotFoundException;
+import br.com.ks.viagem.emvoos.modelos.Companhia;
+import br.com.ks.viagem.emvoos.modelos.Pais;
+import br.com.ks.viagem.emvoos.repositories.CompanhiaRepository;
+import br.com.ks.viagem.emvoos.repositories.PaisRepository;
+import br.com.ks.viagem.emvoos.validators.NomeCompanhiaDuplicadoValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.annotation.InitBinder;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+import java.util.Optional;
+
+import static org.springframework.http.HttpStatus.CREATED;
+
+@RestController
+@RequestMapping("api/v1/companhia/")
+public class CompanhiaController {
+
+    private CompanhiaRepository companhiaRepository;
+    private PaisRepository paisRepository;
+
+    @Autowired
+    public CompanhiaController(CompanhiaRepository companhiaRepository, PaisRepository paisRepository) {
+        this.companhiaRepository = companhiaRepository;
+        this.paisRepository = paisRepository;
+    }
+
+    @InitBinder("companhiaRequest")
+    public void init(WebDataBinder dataBinder) {
+        dataBinder.addValidators(new NomeCompanhiaDuplicadoValidator(companhiaRepository));
+    }
+
+    @ResponseStatus(CREATED)
+    @PostMapping("salvar")
+    public CompanhiaResponse salvarNovaCompanhia(@Valid @RequestBody CompanhiaRequest companhiaRequest) throws Exception {
+        Optional<Pais> pais = paisRepository.findByNome(companhiaRequest.getPais());
```

Comment:
*Avaliamos de forma lógica o nível de complicação de cada trecho de código. Aqui não tem espaço para feeling. Usamos a teoria da carga cognitiva e medimos o nível de pontos de entendimentos necessário por trecho.*

Na minha opinião, devemos manter a carga cognitiva do controller sempre baixa. Já que é o ponto de entrada da requisição e deveria ser o mais fácil possível de entender o fluxo da requisição dentro do sistema.

Nesse caso aqui, adicionamos uma verificação de existência do nome do país, o que até parece uma validação, já que você não quer aceitar nomes de países que não existem. Se você joga a validação para lá, já lima esse if daqui e diminui a carga cognitiva. Eu costumo receber o id e não tratar o fato desse id não existir... Deixo estourar o problema da optional mesmo. 

Por último, coesão, em geral, ajuda a diminuir a carga cognitiva do código, já que você vai deixar junto coisas que são relacionadas. Eu geralmente falo que devemos maximizar o uso do this. Então eu faria o seguinte:

```
Companhia novaCompanhia = companhiaRequest.novaCompanhia(paisRepository);
```
Recebero o repository como argumento pode doer os olhos, mas você cria um caminho claro para construir uma nova companhia em função do formulário.

Reply:
Aqui eu mostro um pouco => https://www.youtube.com/watch?v=PbAYJC96dSI&list=PLVHlvMRWE0Y70QZLWvQvDxxQtFeOEHFkY&index=7

Reply:
Fiz uma alteração no meu entendimento e gostei do resultado. Basicamente quando for criar uma companhia, dentro desse método, existe uma regra para validar a existência do pais, se não existir ele apresente erro. 

Sobre passar o parâmetro do PaisRepository na função, isso me poupa a injeção de um bean dentro do meu CompanhiaRequest, que no meu ponto de vista, fica pior que passar o repository no parametro da função.